File: javascript/0017-letter-combinations-of-a-phone-number.js
/**
 * 17. Letter Combinations of a Phone Number
 * https://leetcode.com/problems/letter-combinations-of-a-phone-number/
 */

/**
 * @param {string} digits
 * @return {string[]}
 */
var letterCombinations = function (digits) {
    let chars = {
        '2': 'abc',
        '3': 'def',
        '4': 'ghi',
        '5': 'jkl',
        '6': 'mno',
        '7': 'pqrs',
        '8': 'tuv',
        '9': 'wxyz',
    }
    let stack = digits.length ? chars[digits[0]].split('') : []

    for (let i = 1; i < digits.length; i++) {
        let newStack = []
        for (let ch of chars[digits[i]])
            for (let st of stack)
                newStack.push(st + ch)
        stack = newStack
    }

    return stack
};

File: swift/0017-letter-combinations-of-a-phone-number.swift
/**
 * 17. Letter Combinations of a Phone Number
 * https://leetcode.com/problems/letter-combinations-of-a-phone-number/
 **/

class Solution {
    func letterCombinations(_ digits: String) -> [String] {
        let keyboard: [Int: [String]] = [
            1: [],
            2: ["a", "b", "c"],
            3: ["d", "e", "f"],
            4: ["g", "h", "i"],
            5: ["j", "k", "l"],
            6: ["m", "n", "o"],
            7: ["p", "q", "r", "s"],
            8: ["t", "u", "v"],
            9: ["w", "x", "y", "z"]
        ]

        var stack: [String] = [""]
        
        for i in Array(digits).map({ Int(String($0))! }) {
            var next: [String] = []
            for row in stack {
                for ch in keyboard[i]! {
                    next.append( row + ch )
                }
            }
            stack = next
        }

        return stack.count <= 1 ? [] : stack
    }
}